packages/test/test-end-to-end-tests/src/test/mockSharedString.ts

```diff
@@ -0,0 +1,262 @@
+		telemetryContext?: ITelemetryContext | undefined,
+		incrementalSummaryContext?: IExperimentalIncrementalSummaryContext | undefined,
+	): ISummaryTreeWithStats {
+		const builder = new SummaryTreeBuilder();
+
+		const content: ISnapshot = {
+			inserts: this.inserts,
+			baseText: this._baseText,
+			currentSequenceNumber: this._sequenceNumber,
+			minimumSequenceNumber: this._minimumSequenceNumber,
+		};
+
+		builder.addBlob(headerBlobName, JSON.stringify(content));
+		return builder.getSummaryTree();
+	}
+	protected async loadCore(storage: IChannelStorageService): Promise<void> {
+		const content = await readAndParse<ISnapshot>(storage, headerBlobName);
+		this._inserts = content.inserts;
+		this._baseText = content.baseText;
+		this._sequenceNumber = content.currentSequenceNumber;
+		this._minimumSequenceNumber = content.minimumSequenceNumber;
+	}
+
+	// Op processing, including stashed ops!
+	protected processCore(
+		message: ISequencedDocumentMessage,
+		local: boolean,
+		localOpMetadata: unknown,
+	) {
+		if (message.type === MessageType.Operation) {
+			const op = message.contents as IInsertOp;
+			switch (op.type) {
+				case "insert": {
+					assert(
+						op.referenceSequenceNumber >= this._minimumSequenceNumber,
+						`Op with ref ${op.referenceSequenceNumber} can't be applied with min ${this._minimumSequenceNumber}!`,
+					);
+
+					if (local) {
+						const insert = this._localInserts.find(
+							(ins) => ins.localIndex === localOpMetadata,
+						);
+						assert(insert !== undefined, "Insert not found");
+						this._localInserts.splice(this._localInserts.indexOf(insert), 1);
+					}
+					const trimmableInsert: ITrimmableInsert = {
+						sequenceNumber: message.sequenceNumber,
+						text: op.text,
+						index: op.index,
+						referenceSequenceNumber: op.referenceSequenceNumber,
+					};
+
+					// Put the insert in the right place to maintain user intent.
+					let i = this.inserts.length;
+					while (
+						i > 0 &&
+						this.inserts[i - 1].referenceSequenceNumber >
+							trimmableInsert.referenceSequenceNumber
+					) {
+						i--;
+					}
+					this.inserts.splice(i, 0, trimmableInsert);
+
+					this.trimLocalState(message.minimumSequenceNumber);
+					this._sequenceNumber = message.sequenceNumber;
+					this._minimumSequenceNumber = message.minimumSequenceNumber;
+					this.emit("valueChanged", trimmableInsert);
+					break;
+				}
+				default:
+					throw new Error("Unknown operation");
+			}
+		}
+	}
+
+	// This trims the global local inserts so that the number of IInserts that need to be processed are based on the minimum sequence number.
+	private trimLocalState(minimumSequenceNumber: number) {
+		while (this.inserts.length > 0 && this.inserts[0].sequenceNumber < minimumSequenceNumber) {
+			const insert = this.inserts.shift();
+			assert(insert !== undefined, "Insert not found");
+			this._baseText = applyInsert(this._baseText, insert);
+		}
+	}
+
+	// Yeah, this should be simple as heck
```

Comment:
You will have to implement reSubmit() - the default implementation of SharedObject will not be sufficient.